doc/ref_arch/kubernetes/chapters/chapter02.md

```diff
@@ -20,20 +20,155 @@
+| `req.inf.ntw.06` | Network | The Architecture **must** support network resiliency. |
+| `req.inf.ntw.07` | Network | The Architecture **should** embrace open-based standards and technologies. |
+| `req.inf.ntw.08` | Network | The Architecture **must** be fully redundant. |
+| `req.inf.ntw.09` | Network | The Architecture **should** support life-cycle management (LCM) of the infrastructure that allows repeatable, scalable and standardised handling of K8s clusters. LCM use cases to cover are: instantiation, termination, upgrade, scaling and healing. |
+| `req.inf.ntw.10` | Network | The networking solution **should** be configurable in an automated manner using openly published API definitions. |
+| `req.inf.ntw.11` | Network | The networking solution **should** be able to be centrally administrated and configured. |
+| `req.inf.ntw.12` | Network | The Architecture **must** support dual stack IPv4 and IPv6 for Kubernetes workloads. |
+| `req.inf.ntw.13` | Network | The Architecture **should** support IPv6 for Kubernetes workloads. |
+| `req.inf.ntw.14` | Network | The Architecture **must** support capabilities for integrating SDN controllers. |
+| `req.inf.ntw.15` | Network | The Architecture **should** support Service Mesh Interface (SMI). |
+| `req.inf.ntw.16` | Network | The Architecture **should**  support co-exisence of multiple Conatiner Network Interface (CNI). |
+| `req.inf.acc.01` | Acceleration | The Architecture **should** support Application Specific Acceleration. |
+| `req.inf.acc.02` | Acceleration | The Architecture **should** support NFVI Acceleration (such as SmartNICs). |
+| `req.inf.vir.01`   | Virtualisation |   The Architecture **must** support the capability for Containers to consume virtualised compute, storage and network resources.|
+| `req.inf.phy.01`  |  Physical Infrastructure |   The Architecture **must** support the capability for Containers to consume physical (aka bare metal) compute, storage and network resources. |
+
+<p align="center"><b>Table 2-2:</b> Kubernetes Architecture: Infrastructure Requirements</p>
+Please note that "shared" is a reference to multi-tenant support and pooled storage resources.
+
+<a name="2.3.3"></a>
+### 2.3.3 Kubernetes Cluster Requirements
+
+| Ref # | sub-category | Description |
+|----|------------------------------------------------|---------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
+| `req.kcm.01` | General | The Architecture **must** allow infrastructure resource sharing within a Kubernetes cluster. |
+| `req.kcm.02` | General | The Architecture **must** support discoverability of nodes and their features. |
+| `req.kcm.03` | General | The Architecture **must** support scheduling of workloads based on Enhanced Platform Awareness (EPA) features such as CPU Pinning, huge-pages and SR-IOV. |
```

Comment:
In RA1 we have the explicit requirement (multiple times): "The Architecture should **not** rely on SR-IOV PCI-Pass through to provide acceleration to VNFs." ... Why we are then mentioning SR-IOV here?

Reply:
Can we remove the references to EPA and let #592 to fix this?

Reply:
Fine for me but propose to delete SR-IOV example as well for now. E.g.:
| `req.kcm.03` | General | The Architecture **must** support scheduling of workloads based on features such as CPU Pinning and huge-pages. |

Reply:
If we would like to build an RA what is compliant with the Network Intensive and Compute Intensive profiles of the RM we must have support for SR-IOV. See nfvi.net.acc.cfg.002 in the RM.

Reply:
@CsatariGergely Thanks for the reminder that we have defined all these features already in the RM within the Profiles. 

Wouldn't that allow to reformulate the requirement to: 
| `req.kcm.03` | General | The Architecture **must** support scheduling of workloads based on defined and supported profiles (Basic, Network Intensive, Compute Intensive). |
?

Concrete features follow then from the profiles as defined in RM. Not sure if we have to list them again. I agree that we need additional a requirement which defines that the RA has to support all three (or less?) profiles.

BTW: nfvi.net.acc.cfg.002 requires only "Support of HW offload" ... with technology examples how to do it. I consider it a mistake that within profile the concrete technologies are required. They should be implementation choices.

That would require that the actual choosen technology should be transparent  for the CNFs and abstracted. Accordingly we have in RA1 the requirement "The Architecture should not rely on SR-IOV PCI-Pass through to provide acceleration to CNFs." (req.int.acc.02, req.inf.acc.03)

Reply:
to quote a meme: "sounds great, doesn't work"
on the cloud infrastructure side sure, there are a couple implementation choices available how to provide direct network access to workloads requiring it.
but they are not that numerous. basically you have 4 choices (SR-IOV, vhostuser stuff from OVS-DPDK, memif stuff from VPP, and maybe, maybe XDP? though haven't seen any production use-cases for the fourth)
all of them require the workload actively knowing about the underlying technology, explicitly requesting the underlying technology, and adapting its internal architecture to use the underlying technology

even for a kernel networking app SR-IOV devices need to be requested through the app's Pod Spec, but for a DPDK app... it is a coupled dependency

so, why the idea behind the requirement is laudable, it is not realistic. the current reality is that the architecture should not rely, but **must support** SR-IOV as an option to accelerate TelCo CNFs

Reply:
@Levovar 
Can you clarify more about "all of them require the workload actively knowing about the underlying technology" ,

Reply:
I was mostly referring to userspace networking aka. DPDK apps
you don't just write an "abstract" DPDK app, it directly depends on the host, and the kernel drivers used on the host
the different networking technologies require the usage of different poll-mode drivers
therefore DPDK apps explicitly request an interface provisioned via a certain a backend. IMO it is unlikely a DPDK app will implement support for all options

but let's ditch DPDK for a second. In Openstack you ask for a vNIC, always* provisioned by OVS. In Kubernetes you ask for a kernel NIC, provisioned via a defined backend
devices, such as SR-IOV VFs are explicitly requested through the Resources API

Reply:
ofc it is always an option to define abstract requirements in spite of reality, but IMO it is better to ask for something which is inline with the real world :)
my only aim with my comments is to somewhat keep this document on the ground, no more, no less

Reply:
Please make sure SR-IOV is **NOT** allowed! this has been discussed in length. SR-IOV MUST NOT BE USED.

Reply:
> you don't just write an "abstract" DPDK app, it directly depends on the host, and the kernel drivers used on the host
> the different networking technologies require the usage of different poll-mode drivers
> therefore DPDK apps explicitly request an interface provisioned via a certain "backend". IMO it is unlikely a DPDK app will implement support for all options

I am not sure this is the only use model of DPDK. DPDK can be used with the application completely abstracted. @petorre can you chip in here?

Reply:
@rabi-abdel please see @levovar and @CsatariGergely comments on SR-IOV.

Reply:
I'm also curious how one writes an app using DPDK, which includes a device specific PMD without depending on the HW, but I'm definitely not a DPDK expert by any means :)
from the stories and use-cases I personally encountered even the PF manufacturer mattered, cause obviously a Mellanox PMD does not have the same capabilities as the Intel one (why would it be simple)

I also assumed we want to share the same NIC between multiple DPDK containers, or even DPDK and non-DPDK containers

anyway: it is one thing to explicitly not mention SR-IOV as an option, but what are your reasons for explicitly forbidding it?
be careful with such strong statements, because they can quickly lead to guaranteed non-compliance on the field

also, if SR-IOV is forbidden, one need to elaborate on how to handle devices, or functionalities which "behave" like a virtual function in the operating system. RDMA through SR-IOV Device Plugin, offload functionalities provided by integrated FPGA come to mind

Reply:
@pgoyal01 could you comment this one out please, I've added to the backlog (#609) as there is an open conversation.

Reply:
It'd make sense to me to tickle out the reasons pro/contra SR/IOV and whether they equally apply to RA1 and RA2. For example:

At the NFVI/VIM layer, operators have to deal with VNFs that were fork-lifted from PNFs and do not support graceful migration of workloads like cloud-native VNFs do. A fallback for operations teams is to use VM live migration. SR/IOV made this cumbersome and kludgy, thus it was frowned upon because it broke the operational model.

In the Kubernetes world, CNFs have to support spinning up new instances on other nodes and migrating/draining workloads off a node. When you do that, SR/IOV becomes much less problematic, but it still gives you the big cost-efficiency gains.

Reply:
In my view the important is the requirement mandating CNI, and we have two: 
req.inf.ntw.01 | Network | The Architecture must support Conatiner Network Interface (CNI).
req.inf.ntw.16 | Network | The Architecture should support co-exisence of multiple Conatiner Network Interface (CNI).
Both have misspelled word "container" and req.inf.ntw.01 is obsolete if we have req.inf.ntw.16.
CNI may consume many different network interfaces including SR-IOV and DPDk and many others now and in the future.  So I agree that we do not have to spell out neither of them in the Requirements.  
What worries me more is that we do not seem to be talking about Multus, which s a container network interface (CNI) plugin that can be used to create multiple network interfaces for pods in Kubernetes, and I think we will need to consider.  Perhaps we do not have it mentioned here in Requirements but we will need this in our RA once we start really defining it.
Regarding comments about SR-IOV and DPDK being a lesser problem for containerised workloads:  SR-IOV and DPDK will still not allow for workload live migration.  It is true that if a workload is a TRUE cloud native  CNF (stateless is critical in this case) then K8s can kill it and spin up a new one somewhere else.  However currently available C/VNFs typically require state persistence, which is problematic with SR-IOV and DPDK.  This can be resolved on the MANO (NFVO) level by a suitable orchestration.  The latter is the topic that (even outside NFVI scope) will need to be addressed by RA2.

Reply:
We have a requirement to have several network interfaces and a requirement to support CNI-s this mean that we will need to discuss CNI metaplugins, like [Multus](https://github.com/intel/multus-cni) or [DANM](https://github.com/nokia/danm).

I'm happy to start to work on this.

Reply:
FYI and full disclosure: I'm one the DANM authors and core maintainer (too :) ).
Always happy to talk about multiple interfaces in K8s with anyone interested!

Reply:
Hi - can we move this conversation to the backlog issue #609 please so we can resolve this PR and merge it. Thanks!

Reply:
I generally agree with your comments: SR-IOV is undesired for many reasons (features, validation, operational, security) but for performance reasons is how huge majority of virtualized dataplane works today.

> 
> I'm also curious how one writes an app using DPDK, which includes a device specific PMD without depending on the HW, but I'm definitely not a DPDK expert by any means :)
> from the stories and use-cases I personally encountered even the PF manufacturer mattered, cause obviously a Mellanox PMD does not have the same capabilities as the Intel one (why would it be simple)

DPDK apps use PMDriver to talk to network device. Currently available drivers are vendor-specific and more generic ones (like virtio, where we still need to mature/optimize such implementations).

> I also assumed we want to share the same NIC between multiple DPDK containers, or even DPDK and non-DPDK containers

Need to be specific with terminology like "NIC" which can consist of various chips from Ethernet Controller on or expose multiple ports. PMD will take assigned device for DPDK app and won't share that device with non-DPDK apps. Carefully writing such requirement will determine what RI would be compliant.

> anyway: it is one thing to explicitly not mention SR-IOV as an option, but what are your reasons for explicitly forbidding it?
> be careful with such strong statements, because they can quickly lead to guaranteed non-compliance on the field
> 
> also, if SR-IOV is forbidden, one need to elaborate on how to handle devices, or functionalities which "behave" like a virtual function in the operating system. RDMA through SR-IOV Device Plugin, offload functionalities provided by integrated FPGA use-cases come to mind

Reply:
@petorre: I would like to echo @tomkivlin:

> Hi - can we move this conversation to the backlog issue #609 please so we can resolve this PR and merge it. Thanks!

This conversation is for this reason already marked as resolved.

Reply:
> If we would like to build an RA what is compliant with the Network Intensive and Compute Intensive profiles of the RM we must have support for SR-IOV. See nfvi.net.acc.cfg.002 in the RM.

@CsatariGergely 
Just small comment here before we move the conversation to another issue
first ( again ) it is a good catch, as we initially have a rule of thumb that we will not cover SRIOV in NFVi profiling ( old name :)) CNTT ( now ) , so we should be removed in this nfvi.net.acc.cfg.002

and regarding your point "if we would like to build an RA what is compliant with the Network Intensive and Compute Intensive profiles of the RM we must have support for SR-IOV"
I am not with you in this , for Compute Intensive profiles can be achieved by Just OVS 
and for network-intensive that required high throughput  it can be achieved by OVS+DPDK 
( For sure we need to check PPS ) 

So I am totally with to avoid SRIOV in all our RM/RAx 

@rabi-abdel @rabiabdel your comment here